### PR TITLE
Storages: Fix comparing null in MinMaxIndex

### LINKED ISSUE #9374: Incorrect result is returned  when comparing with null
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
 drop table if exists test.t;

 create table test.t (a date);

 insert into test.t values("2024-08-26"),("2024-08-25"),("2024-08-24"),("2024-08-23");

 alter table test.t set tiflash replica 1;

 alter table test.t compact tiflash replica;

 set session tidb_isolation_read_engines='tiflash'; select * from test.t where not (a between NULL and '2024-08-25');
```

### 2. What did you expect to see? (Required)
```
+------------+
| a          |
+------------+
| 2024-08-26 |
+------------+
```

### 3. What did you see instead (Required)
```
+------------+
| a          |
+------------+
| 2024-08-26 |
| 2024-08-25 |
| 2024-08-24 |
| 2024-08-23 |
+------------+
```

### 4. What is your TiFlash version? (Required)

f0a4087d599a19645820b0bc9c48d9b80906541f

### PR DESCRIPTION
**Pull request by JinheLin**
### What problem does this PR solve?

Issue Number: close #9374

Problem Summary:

The filter of `select * from test.t where not (a between NULL and '2024-08-25')` will be transformed in TiFlash as `not (a between NULL and '2024-08-25')` => `not (a >= NULL and a <= '2024-08-25')` => `a < NULL or a > '2024-08-25'`.

The problem is how `RSOperator` and `MinMaxIndex` handle `a < NULL`. 

As the code below, `Less` will be transformed into `not GreaterEqual`:
https://github.com/pingcap/tiflash/blob/81cd94782a912d3a61e38b7ebc5c2cfabe5694e9/dbms/src/Storages/DeltaMerge/Filter/Less.h#L32-L37

`MinMaxIndex` execute `GreaterEqual` and the result for comparing with `NULL` is `RSResult::None`.
https://github.com/pingcap/tiflash/blob/81cd94782a912d3a61e38b7ebc5c2cfabe5694e9/dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp#L422-L426

`not RSResult::None` equals to `RSResult::All`. And PR https://github.com/pingcap/tiflash/pull/9361 will skip filter if the filter result of a Block is `RSResult::All`.

**The root cause is that the result of comparing any value with null is null, returning `RSResult::None` is incorrect.**

### What is changed and how it works?

```commit-message
In `MinMaxIndex`, if the input value is null, return `RSResult::NoneNull`.
```

Handling null input value, in `checkCmp`: https://github.com/pingcap/tiflash/blob/81cd94782a912d3a61e38b7ebc5c2cfabe5694e9/dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp#L422-L426

Handling null input value, in `checkIn`: https://github.com/pingcap/tiflash/blob/81cd94782a912d3a61e38b7ebc5c2cfabe5694e9/dbms/src/Storages/DeltaMerge/Index/RoughCheck.h#L76-L78

For some filters like `checkCmp(null)`, the result is `RSResult::NoneNull` and the block can be skipped.
For some filters like `not checkCmp(null)`, the result is `RSResult::AllNull` and the block will be read and filtered.

In fact, the result `checkCmp(null)` and `not checkCmp(null)`  should always be null. Maybe we can extend `RSResult` to support this. However, **comparing(>, <, >=, <=, ==, !=, in) with null is not a normal operation in MySQL. So just let it fallback to `FilterTransforamAction` if neccessary**. 

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by Lloyd-Pottiger — APPROVED** [review]

File: dbms/src/Storages/DeltaMerge/Index/RoughCheck.h [hunk 1]
```diff
@@ -73,10 +73,11 @@ struct CheckIn
         {
             if (result == RSResult::All)
                 break;
-            // skip null value
+
             if (v.isNull())
-                continue;
-            result = result || CheckEqual::check<T>(v, type, min, max);
+                result = result || RSResult::NoneNull;
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/Index/RoughCheck.h:78]:
```suggestion
                result = result || RSResult::NoneNull; // the result of any arithmetic comparison with NULL is also NULL
```

File: dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp [hunk 2]
```diff
@@ -421,9 +419,8 @@ RSResults MinMaxIndex::checkCmpImpl(size_t start_pack, size_t pack_count, const
 template <typename Op>
 RSResults MinMaxIndex::checkCmp(size_t start_pack, size_t pack_count, const Field & value, const DataTypePtr & type)
 {
-    RSResults results(pack_count, RSResult::None);
     if (value.isNull())
-        return results;
+        return RSResults(pack_count, RSResult::NoneNull);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp:423]:
```suggestion
        return RSResults(pack_count, RSResult::NoneNull); // the result of any arithmetic comparison with NULL is also NULL
```

File: dbms/src/Storages/DeltaMerge/tests/gtest_dm_minmax_index.cpp [hunk 3]
```diff
@@ -2662,4 +2662,82 @@ try
...
+        for (auto actual_res : results)
+            ASSERT_EQ(actual_res, expected_res) << fmt::format("actual: {}, expected: {}", actual_res, expected_res);
+    };
+
+    check(createGreater(attr("Int64"), null_value), RSResult::NoneNull);
+    check(createGreaterEqual(attr("Int64"), null_value), RSResult::NoneNull);
+    check(createLess(attr("Int64"), null_value), RSResult::AllNull);
+    check(createLessEqual(attr("Int64"), null_value), RSResult::AllNull);
+    check(createEqual(attr("Int64"), null_value), RSResult::NoneNull);
+    check(createNotEqual(attr("Int64"), null_value), RSResult::AllNull);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/tests/gtest_dm_minmax_index.cpp:2689]: why "less"/"lessequal"/"noteuqal" return "AllNull" but others return "NoneNull"?

File: tests/fullstack-test-dt/expr/compare_null.test
Comment (JaySon-Huang) [inline tests/fullstack-test-dt/expr/compare_null.test:1]: Better add some test case about this: https://dev.mysql.com/doc/refman/8.4/en/working-with-null.html

> Two NULL values are regarded as equal in a GROUP BY.
> When doing an ORDER BY, NULL values are presented first if you do ORDER BY ... ASC and last if you do ORDER BY ... DESC.

**Review by JaySon-Huang — APPROVED** [review]
LGTM

File: tests/fullstack-test-dt/expr/compare_null.test [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+
+mysql> insert into test.t values('2024-08-21', 6),('2024-08-22', 7),('2024-08-23', 8),('2024-08-24', 9), (NULL, 10);
+
+mysql> alter table test.t set tiflash replica 1;
+
+func> wait_table test t
+
+mysql> alter table test.t compact tiflash replica;
+
+mysql> set session tidb_isolation_read_engines='tiflash'; select sum(b) from test.t group by a order by a desc;
```
Comment (JaySon-Huang) [inline tests/fullstack-test-dt/expr/compare_null.test:234]:
```suggestion
mysql> set session tidb_isolation_read_engines='tiflash'; select a, sum(b) from test.t group by a order by a desc;
```

File: tests/fullstack-test-dt/expr/compare_null.test [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+| a          | sum(b) |
++------------+--------+
+| 2024-08-24 |     13 |
+| 2024-08-23 |     11 |
+| 2024-08-22 |      9 |
+| 2024-08-21 |      7 |
+| NULL       |     15 |
++------------+--------+
+
+mysql> set session tidb_isolation_read_engines='tiflash'; select sum(b) from test.t group by a order by a desc;
```
Comment (JaySon-Huang) [inline tests/fullstack-test-dt/expr/compare_null.test:245]:
```suggestion
mysql> set session tidb_isolation_read_engines='tiflash'; select a, sum(b) from test.t group by a order by a desc;
```

**Pull request merged**